finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4DispatcherBase.scala

```diff
@@ -42,6 +50,33 @@ trait Netty4DispatcherBase[SendMsg <: Message, RecvMsg <: Message] {
     streams.size()
   }
 
+  @volatile private[this] var pingFramePromise: Promise[Unit] = null
```

Comment:
Consider making this an AtomicReference and using cas state transitions to avoid race conditions.